.github/workflows/python-packaging.yml

```diff
@@ -1,35 +1,37 @@
+# Workflow to build distribution, and push to PyPi or TestPyPi
+
 name: TileDB-SOMA python sdist & wheels
 
 on:
-  # I.e. on request
+  # Trigger publication to TestPyPi via user workflow request:
+  #     gh workflow run python-packaging.yml --ref branch
+  # Specify the branch or other ref as required.
   workflow_dispatch:
-  # Normally the smoke test only runs on merge to main.  To run it on a PR CI, uncomment `pull_request:`
-  # pull_request:
-  #   branches: [main]
-  # paths-ignore:
-  #   - 'apis/r/**'
+
+  # Trigger publication to PyPi via new release event.
   release:
     types: [published]
-  schedule:
-    - cron: "42 9 * * *"
```

Comment:
Thanks for getting to this, looks great, my only question is whether to keep the nightly build? It was a way of warning us about package-breaking changes that inadvertently make it onto main (since the packaging workflow doesn't otherwise trigger automatically + routinely). Which has happened a number of times.

Reply:
There is a discussion in slack about this point.

TL;DR:
* PyPi/TestPyPi won't accept the same version number twice
* to do a nightly publication, we would need to only do it when there has been a commit, i.e., a new version number will be generated.
* So as originally written, the workflow would fail on any day when there was not a commit to main
* I volunteered to add this gate (ie. skip if no commit in the past 24 hours) if it was desired, and as an interim, removed the scheduled run.

So LMK if this is worth adding.  I'll probably do it as a separate PR, as it makes the TestPyPi action logic more complex.

Reply:
Since we don't care about version ordering on testpypi, we could do something like this:

1. Generate a fake version number using the commit hash (e.g. 0.0.[convert 32 bits of the commit hash to an integer])
2. See if that release already exists in testpypi.
3. If not, make a release to testpypi.

Reply:
That would require a different build process, right?  The version is encoded in the whl/sdist, which is shared by both workflows.  Are you aware of a means of revising the version during the twine/upload process?

Reply:
If it's a real workflow, use the normal "real version" logic; if it's a fake workflow (nightly / manual), override the entire thing with a fake version.

Reply:
I"m not following - I'll need to touch base offline to get further details, if we want to proceed down this path.  Let me know if that is the case and I"ll hit you in another venue.

Reply:
If auto-pushing to TestPyPI causes complications, I think it'd be fine to *not* push to it at all for the nightly build. I believe this would be a condition something like `if: ${{ github.event_name != 'schedule' }}`

There is still some value in frequently exercising the preceding parts of this workflow that build the wheels and run basic checks on them. (Yet it's not suitable to do so for every push to main, since it uses a lot of CI resources to fill in the platform/pyversion matrix)

Reply:
to be clear, my suggestions here were not something to do on this change, but thoughts for the future if we ever want to go in that direction.

Reply:
OK, I am now understanding!

I'll make an amendment to the PR along the lines Mike has suggested and re-tag you for review.  I.e., execute the workflow on a cron, _excluding_ TestPyPi/PyPi push.

Thanks!

Reply:
OK, PTAL as of `9bee92c`:
- added the scheduled run
- TestPyPi is conditional upon user request event
- PyPi is conditional on release event
- modified the error handler (bug file) to be dependent on all branches so it will catch failure in any invocation, including the scheduled event

Reply:
I also revalidated the TestPyPi functionality.  [The run is here](https://github.com/single-cell-data/TileDB-SOMA/actions/runs/5015669731).